---
title: "GEMMA"
output: html_document
---

**Last updated:** `r Sys.Date()`

**Code version:** `r git2r::commits()[[1]]@sha`

```{r knitr, include=FALSE}
library("knitr")
opts_chunk$set(fig.width = 10, fig.path = "figure/")
```

Results from GEMMA.

## Setup

```{r packages, message=FALSE}
library("cowplot")
library("ggplot2")
library("magrittr")
library("plyr"); library("dplyr")
library("stringr")
library("tidyr")
theme_set(theme_cowplot(font_size = 12))
```

Input GEMMA results.

```{r input-gemma}
f_gemma <- Sys.glob("../data/gemma/top-pca-*.txt")
l_gemma <- list()
for (f in f_gemma) {
  n <- f %>% basename %>% str_replace("top-", "") %>% str_replace(".txt", "")
  # if (n %in% c("pca-40")) next
  l_gemma[[n]] <- read.delim(f, stringsAsFactors = FALSE)
}
d_gemma <- ldply(l_gemma, .id = "analysis")
d_gemma$analysis <- factor(d_gemma$analysis,
                           levels = paste0("pca-", c(0:10, 15, 20, 30, 40)))
summary(d_gemma)
```

## Initial inspection

For the intial inspection, juse use the base case of zero PCs as covariates.

```{r gemma-qc}
gemma <- d_gemma %>% filter(analysis == "pca-0")
p_n_snps <- ggplot(gemma, aes(x = n_snps)) +
  geom_histogram() +
  labs(x = "Number of SNPs tested",
       y = "Number of genes",
       title = "Number of SNPs\ntested per gene")
p_n_chr <- ggplot(gemma, aes(x = chr)) +
  geom_bar() +
  scale_x_continuous(breaks = seq(1, max(gemma$chr), 2)) +
  labs(x = "Autosomes",
       y = "Number of genes tested",
       title = "Number of genes\ntested per chromosome")
plot_grid(p_n_snps, p_n_chr, nrow = 1)
```

The following metrics will be slightly different across the PC analyses because they do not always find the same top SNP, but the pca-0 case should be representative.

```{r gemma-miss-af-mle}
p_n_miss <- ggplot(gemma, aes(x = n_miss)) +
  scale_x_continuous(breaks = seq(0, max(gemma$n_miss))) +
  geom_histogram() +
    labs(x = "Number of missing individuals",
       y = "Number of genes",
       title = "Number of missing individuals\nper eQTL-gene test")
p_af <- ggplot(gemma, aes(x = af)) +
  scale_x_continuous(breaks = seq(0, max(gemma$af), by = 0.1)) +
  geom_histogram() +
    labs(x = "Allele Frequency",
       y = "Number of eQTLs (1 per gene)",
       title = "Allele frequency distribution of eQTLs")
plot_grid(p_n_miss, p_af)
```

The metric "l_mle" is the maximum likelihood estimate for $\lambda$. From the manual:

> The algorithm calculates either REML or MLE estimate of $\lambda$ in the evaluation
> interval from 1×10^−5^ (corresponding to almost pure environmental effect) to
> 1×10^5^ (corresponding to almost pure genetic effect).

Across all the analyses, $\lambda$ is always estimated to be 1x10^5^.

```{r l_mle}
unique(d_gemma$l_mle)
```

## Multiple testing correction

Bonferroni per gene.

```{r bonferroni}
bonferroni <- function(p, n) {
  # p - nominal p-value(s)
  # n - number(s) of tests
  stopifnot(is.numeric(p), p >= 0, p <= 1,
            is.numeric(n), n > 0)
  n_valid_len <- length(n) == 1 || length(n) == length(p)
  if (!n_valid_len)
    stop("n must be length 1 or the same length as p")
  p_b <- pmin(1, p * n)
  stopifnot(p_b <= 1)
  return(p_b)
}

d_gemma$p_bonf <- bonferroni(d_gemma$p_lrt, d_gemma$n_snps)
```

Benjamini-Hochberg across eQTLs.

```{r bh}
d_gemma <- d_gemma %>%
  group_by(analysis) %>%
  mutate(p_bh = p.adjust(p_bonf, method = "BH")) %>%
  ungroup()
```

Inspecting the p-value distributions.

```{r gemma-p-values, fig.width=12, fig.height=6}
d_gemma_p <- d_gemma %>%
  select(analysis, n_pcs, chr, starts_with("p_")) %>%
  gather("type", "p", starts_with("p_"))
d_gemma_p$type <- factor(d_gemma_p$type,
                         levels = c("p_lrt", "p_bonf", "p_bh"),
                         labels = c("Nominal", "Bonferroni per gene",
                                    "BH across genes"))
p_pval <- ggplot(d_gemma_p, aes(x = p)) +
  geom_histogram() +
  facet_grid(type ~ analysis, scales = "free_y") +
  scale_x_continuous(breaks = seq(0, 1, 0.25),
                     limits = c(-0.1, 1.1)) +
  theme(axis.text.x = element_text(angle = -45, vjust = 0.01)) +
  labs(x = "p-value",
       y = "Number of eQTLs",
       title = "p-values distributions")
p_pval
```

## Significant eQTLs

Optimize at 10% FDR.

```{r gemma-sig}
d_gemma_sig <- d_gemma %>%
  group_by(analysis, n_pcs) %>%
  summarize(sig = sum(p_bh < 0.10))
p_sig <- ggplot(d_gemma_sig, aes(x = n_pcs, y = sig)) +
  geom_point() +
  scale_x_continuous(breaks = seq(min(d_gemma_sig$n_pcs),
                                  max(d_gemma$n_pcs), by = 1)) +
  labs(x = "Number of PCs included as covariates",
       y = "Number of significant eQTLs at 10% FDR",
       title = "Effect of PCs on number of eQTLs detected")
p_sig
```

## Incorporate TSS

Input TSS information.

```{r input-tss}
tss <- read.table("../data/tss.txt", stringsAsFactors = FALSE)
colnames(tss) <- c("gene", "chr", "tss", "strand", "biotype")
head(tss)
```

Merge with GEMMA results.

```{r merge-tss}
d_gemma_tss <- merge(d_gemma, tss)
```

Calculate distance between eQTL and TSS.

```{r eQTL-distance}
d_gemma_tss$dist <- ifelse(d_gemma_tss$strand == "+",
                           d_gemma_tss$ps - d_gemma_tss$tss,
                           -(d_gemma_tss$ps - d_gemma_tss$tss))
d_gemma_tss$sig <- d_gemma_tss$p_bh < 0.10
```

Plot.

```{r dist-density}
p_dist <- ggplot(d_gemma_tss %>% filter(analysis != "pca-40"),
                 aes(x = dist, color = sig)) +
  geom_density() +
  facet_wrap(~analysis, ncol = 5) +
  scale_x_continuous(name = "Distance from TSS (upstream <- -> downstream)",
                     breaks = seq(-10^6, 10^6, by = 50 * 10^4),
                     labels = c("-1 Mb", "-500 kb", "TSS", "+500 kb", "+1 Mb"),
                     limits = c(-10^6, 10^6)) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0.5, vjust = 0.75)) +
  scale_color_discrete(name = "", labels = c("non-eQTL", "eQTL at 10% FDR"))
p_dist
```

```{r dist-hist}
p_dist_hist <- ggplot(d_gemma_tss, aes(x = dist, fill = sig)) +
  geom_histogram() +
  facet_wrap(~analysis, ncol = 5) +
  scale_x_continuous(name = "Distance from TSS (upstream <- -> downstream)",
                     breaks = seq(-10^6, 10^6, by = 50 * 10^4),
                     labels = c("-1 Mb", "-500 kb", "TSS", "+500 kb", "+1 Mb"),
                     limits = c(-10^6, 10^6)) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0.5, vjust = 0.75)) +
  scale_fill_discrete(name = "", labels = c("non-eQTL", "eQTL at 10% FDR"))
p_dist_hist
```

```{r dist-by-pc}
p_dist_pc <- ggplot(d_gemma_tss %>% filter(analysis != "pca-40"),
                    aes(x = dist, color = analysis)) +
  geom_density() +
  facet_wrap(~sig) +
  scale_x_continuous(name = "Distance from TSS (upstream <- -> downstream)",
                     breaks = seq(-10^6, 10^6, by = 50 * 10^4),
                     labels = c("-1 Mb", "-500 kb", "TSS", "+500 kb", "+1 Mb"),
                     limits = c(-10^6, 10^6)) +
  theme(axis.text.x = element_text(angle = -45, hjust = 0.5, vjust = 0.75))
p_dist_pc
```

## Session information

```{r session-info}
sessionInfo()
```

