---
title: "GEMMA"
output: html_document
---

**Last updated:** `r Sys.Date()`

**Code version:** `r git2r::commits()[[1]]@sha`

```{r knitr, include=FALSE}
library("knitr")
opts_chunk$set(fig.width = 10, fig.path = "figure/")
```

Results from GEMMA.

## Setup

```{r packages, message=FALSE}
library("cowplot")
library("dplyr")
library("ggplot2")
theme_set(theme_cowplot(font_size = 12))
```

Input Subread counts.

```{r}
gemma <- read.delim("../data/gemma/gemma-top.txt", stringsAsFactors = FALSE)
dim(gemma)
head(gemma)
```

## Initial inspection

```{r gemma-qc}
p_n_snps <- ggplot(gemma, aes(x = n_snps)) +
  geom_histogram() +
  labs(x = "Number of SNPs tested",
       y = "Number of genes",
       title = "Number of SNPs\ntested per gene")
p_n_miss <- ggplot(gemma, aes(x = n_miss)) +
  scale_x_continuous(breaks = seq(0, max(gemma$n_miss))) +
  geom_histogram() +
    labs(x = "Number of missing individuals",
       y = "Number of genes",
       title = "Number of missing individuals\nper eQTL-gene test")
p_n_chr <- ggplot(gemma, aes(x = chr)) +
  geom_bar() +
  scale_x_continuous(breaks = seq(1, max(gemma$chr), 2)) +
  labs(x = "Autosomes",
       y = "Number of genes tested",
       title = "Number of genes\ntested per chromosome")
plot_grid(p_n_snps, p_n_miss, p_n_chr, nrow = 1)
```

## Multiple testing correction

Bonferroni per gene.

```{r bonferroni}
bonferroni <- function(p, n) {
  # p - nominal p-value(s)
  # n - number(s) of tests
  stopifnot(is.numeric(p), p >= 0, p <= 1,
            is.numeric(n), n > 0)
  n_valid_len <- length(n) == 1 || length(n) == length(p)
  if (!n_valid_len)
    stop("n must be length 1 or the same length as p")
  p_b <- pmin(1, p * n)
  stopifnot(p_b <= 1)
  return(p_b)
}

gemma$p_bonf <- bonferroni(gemma$p_lrt, gemma$n_snps)
```

Benjamini-Hochberg across eQTLs.

```{r bh}
gemma$p_bh <- p.adjust(gemma$p_bonf, method = "BH")
```

```{r gemma-p-values}
p_p_lrt <- ggplot(gemma, aes(x = p_lrt)) +
  geom_histogram() +
  scale_x_continuous(breaks = seq(0, 1, 0.25),
                     limits = c(-0.1, 1.1)) +
  labs(x = "p-value",
       y = "Number of eQTLs",
       title = "p-values from\nlikelihood ratio test")
p_p_bonf <- ggplot(gemma, aes(x = p_bonf)) +
  geom_histogram() +
  scale_x_continuous(breaks = seq(0, 1, 0.25),
                     limits = c(-0.1, 1.1)) +
  labs(x = "p-value",
       y = "Number of eQTLs",
       title = "Bonferroni adjusted p-values\nwithin each gene")
p_p_bh <- ggplot(gemma, aes(x = p_bh)) +
  geom_histogram() +
  scale_x_continuous(breaks = seq(0, 1, 0.25),
                     limits = c(-0.1, 1.1)) +
  labs(x = "p-value",
       y = "Number of eQTLs",
       title = "BH adjusted p-values\nacross eQTLs")
plot_grid(p_p_lrt, p_p_bonf, p_p_bh, nrow = 1)
```

## Significant eQTLs

```{r}
sum(gemma$p_bh < 0.05)
sum(gemma$p_bh < 0.1)
sum(gemma$p_bh < 0.2)
table(gemma$p_bh < 0.05, gemma$chr)
```

```{r gemma-per-chr}
gemma_chr <- gemma %>%
  group_by(chr) %>%
  summarize(n_total = n(),
            n_sig = sum(p_bh < 0.05),
            n_stand = n_sig / n_total)
p_n_sig <- ggplot(gemma_chr, aes(x = chr, y = n_sig)) +
  geom_bar(stat = "identity") +
  scale_x_continuous(breaks = seq(1, max(gemma_chr$chr), 2)) +
  scale_y_continuous(breaks = seq(0, max(gemma_chr$n_sig), 2)) +
  labs(x = "Autosomes",
       y = "Number of eQTLs with p < 0.05",
       title = "Significant eQTLs per autosome")
p_n_stand <- ggplot(gemma_chr, aes(x = chr, y = n_stand)) +
  geom_bar(stat = "identity") +
  scale_x_continuous(breaks = seq(1, max(gemma_chr$chr), 2)) +
  # scale_y_continuous(breaks = seq(0, max(gemma_chr$n_sig), 2)) +
  labs(x = "Autosomes",
       y = "Number of eQTLs with p < 0.05 / Number of genes",
       title = "Significant eQTLs per autosome\nstandardized by number of genes tested")
plot_grid(p_n_sig, p_n_stand, nrow = 1)
```

```{r gemma-fdr-cutoff}
cutoffs <- seq(0, 0.5, by = 0.05)
n_sig <- numeric(length = length(cutoffs))
for (i in seq_along(cutoffs)) {
  n_sig[i] <- sum(gemma$p_bh < cutoffs[i])
}
d_cutoffs <- data.frame(cutoffs, n_sig)
ggplot(d_cutoffs, aes(x = cutoffs, y = n_sig)) +
  geom_point() +
  scale_x_continuous(breaks = cutoffs) +
  labs(x = "BH FDR cutoff",
       y = "Number of significant eQTLs",
       title = "Effect of FDR cutoff")
```


## Session information

```{r session-info}
sessionInfo()
```

